cmd/db-exporter/src/main.rs

```diff
@@ -1483,3 +1536,125 @@ pub fn apply_snapshot(
     println!("apply snapshot use time: {:?}", use_time.as_secs());
     Ok(())
 }
+
+#[derive(Serialize, Debug)]
+pub struct AccountData<R: Serialize> {
+    address: AccountAddress,
+    #[serde(flatten)]
+    resource: Option<R>,
+}
+
+pub fn export_resource(
+    db: &str,
+    output: &Path,
+    block_id: HashValue,
+    resource_struct_tag: StructTag,
+    fields: &[String],
+) -> anyhow::Result<()> {
+    let db_storage = DBStorage::open_with_cfs(
+        db,
+        StorageVersion::current_version()
+            .get_column_family_names()
+            .to_vec(),
+        true,
+        Default::default(),
+        None,
+    )?;
+    let storage = Storage::new(StorageInstance::new_db_instance(db_storage))?;
+    let storage = Arc::new(storage);
+    let block = storage
+        .get_block(block_id)?
+        .ok_or_else(|| anyhow::anyhow!("block {} not exist", block_id))?;
+
+    let root = block.header.state_root();
+    let statedb = ChainStateDB::new(storage, Some(root));
+    let value_annotator = MoveValueAnnotator::new(&statedb);
+
+    // let state_tree = StateTree::<AccountAddress>::new(storage.clone(), Some(root));
+
+    let mut csv_writer = csv::WriterBuilder::new().from_path(output)?;
+
+    // write csv header.
+    {
+        csv_writer.write_field("address")?;
+        for f in fields {
+            csv_writer.write_field(f)?;
+        }
+        csv_writer.write_record(None::<&[u8]>)?;
+    }
+
+    let global_states_iter = statedb.dump_iter()?;
+    for (account_address, account_state_set) in global_states_iter {
```

Comment:
time inside iterator loop:
original code took: 1267ms
this version takes: 5525ms

Reply:
what's the block height? could you help me to analyze the reason? If you are using a linux server, you can use the flamegraph see the difference

Reply:
height is 6,881,689

Reply:
https://github.com/starcoinorg/starcoin/blob/6c0546e1326387d27b6c92cb3c449301dac229b9/cmd/db-exporter/src/main.rs#L458-L466

this is the related code

Reply:
![flamegraph](https://user-images.githubusercontent.com/23182033/181143698-75fa8703-1686-4ef8-8e2a-20a3010bf3af.svg)

Reply:
flamegraph of original `resource-exporter`
![flamegraph-resource-exporter](https://user-images.githubusercontent.com/23182033/181146942-443dbcb2-726c-4109-8b15-d955aefea9dc.svg)

Reply:
can you compare new commit and the original resource_exporter flamegraph

Reply:
No clue yet, trying👀

Reply:
thx, I want to know reason. Let me think.

Reply:
just measured time as in 3808f98
turns out the iterator (`global_states_iter.next()`) took most of time, about 5000ms

The function corresponding to `global_states_iter.next` in flamegraph should be
`<starcoin_state_tree::state_tree::AccountStateSetIterator as core::iter::traits::iterator::Iterator>::next
`, inside which `rocksdb::DBImplReadOnly::Get(rocksdb::ReadOptions const&, rocksdb::ColumnFamilyHandle*, rocksdb::Slice const&, rocksdb::PinnableSlice*)` account for ~96% of time